src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs

```diff
@@ -146,28 +156,55 @@ protected async Task WaitMessageIntervalDuringRecoveryAsync(CancellationToken ca
             Logger.LogDebug("Circuit breaker caused message pump '{JobId}' to wait message interval during recovery of '{Interval}' during the '{State}' state", JobId, CircuitState.Options.MessageIntervalDuringRecovery.ToString("g"), CircuitState);
             await Task.Delay(CircuitState.Options.MessageIntervalDuringRecovery, cancellationToken);
 
+            MessagePumpCircuitState oldState = CircuitState;
+            bool shouldTriggerTransitionEvent = !oldState.IsHalfOpen;
+
             CircuitState = CircuitState.TransitionTo(CircuitBreakerState.HalfOpen);
+
+            if (shouldTriggerTransitionEvent)
+            {
+                await NotifyCircuitBreakerEventHandlersAsync();
+            }
         }
 
         /// <summary>
         /// Notifies the message pump about the new state which pauses message retrieval.
         /// </summary>
         /// <param name="options">The additional accompanied options that goes with the new state.</param>
-        internal void NotifyPauseReceiveMessages(MessagePumpCircuitBreakerOptions options)
+        internal async Task NotifyPauseReceiveMessagesAsync(MessagePumpCircuitBreakerOptions options)
         {
             Logger.LogDebug("Circuit breaker caused message pump '{JobId}' to transition from a '{CurrentState}' an 'Open' state", JobId, CircuitState);
 
             CircuitState = CircuitState.TransitionTo(CircuitBreakerState.Open, options);
+
+            await NotifyCircuitBreakerEventHandlersAsync();
         }
 
         /// <summary>
         /// Notifies the message pump about the new state which resumes message retrieval.
         /// </summary>
-        protected void NotifyResumeRetrievingMessages()
+        protected async Task NotifyResumeRetrievingMessagesAsync()
         {
             Logger.LogDebug("Circuit breaker caused message pump '{JobId}' to transition back from '{CurrentState}' to a 'Closed' state, retrieving messages is resumed", JobId, CircuitState);
 
             CircuitState = MessagePumpCircuitState.Closed;
+
+            await NotifyCircuitBreakerEventHandlersAsync();
+        }
+
+        private async Task NotifyCircuitBreakerEventHandlersAsync()
+        {
+            ICircuitBreakerEventHandler[] eventHandlers = GetEventHandlersForPump();
+
+            await Task.WhenAll(eventHandlers.Select(h => h.OnTransitionAsync(CircuitState)));
```

Comment:
Was wondering if this is a good place to use `Task.WhenAny` instead 🤔

Reply:
There's a difference between the 2; `WhenAll` waits 'till all tasks are completed, `WhenAny` stops waiting as soon as one task has completed.
I wonder why `WaitAny` would be appropriate ?

Reply:
In hindsight ....  indeed, is it necessary to await on those eventhandlers ?  Is it even necessary to await on any of them ?  Isn't this just 'fire and forget' cfr events in Windows applications ?